feat: Button 컴포넌트 완료

KimChunsick opened the pull request:
feat: Button 컴포넌트 작업 완료 및 blog에서 확인 가능하게 컴포넌트 추가

Signed-off-by: deokwon <deokwon@deokwonui-MacBookPro.local>

Commit:
feat: Button 컴포넌트 완료

feat: Button 컴포넌트 작업 완료 및 blog에서 확인 가능하게 컴포넌트 추가

Signed-off-by: deokwon <deokwon@deokwonui-MacBookPro.local>

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,118 @@
+    css`
+      height: ${props.size === 'small' ? '32px' : props.size === 'medium' ? '40px' : '48px'};
+    `}
+    box-shadow: ${props.variant === 'shadow' ? '0 5px 10px rgba(0,0,0,0.12)' : undefined};
```
props.theme.shadows 가 있음.

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,118 @@
+      background: ${props.hover?.background};
+      border: 1px solid ${props.hover?.border};
+      transform: ${props.variant === 'shadow' ? 'translateY(-2px)' : undefined};
+      box-shadow: ${props.variant === 'shadow' ? '0 8px 30px rgba(0,0,0,0.12)' : undefined};
```
props.theme.shadows 가 있음.

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,118 @@
+      background: ${props.active?.background};
+      border: 1px solid ${props.active?.border};
+      transform: ${props.variant === 'shadow' ? 'none' : undefined};
+      box-shadow: ${props.variant === 'shadow' ? '0 5px 10px rgba(0,0,0,0.12)' : undefined};
```
props.theme.shadows 가 있음.

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,118 @@
+
+  &:disabled {
+    cursor: not-allowed;
+    color: ${Colors.PRIMARY.ACCENT_3};
```
이렇게 쓰는 거 아님.

```tsx
css`
color: ${({ theme }) => theme.colors.PRIMARY.ACCENT_3};
`
```

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,118 @@
+
+const Accessory = styled.span<{ isPrefix?: boolean }>`
+  ${(props) => css`
+    margin-right: ${props.isPrefix ? '8px' : undefined};
```
```tsx
props.isPrefix && css`
margin-right: 8px; 
`
```

KimChunsick replied:
허허.. 죄송함다..

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,118 @@
+  overflow: hidden;
+  display: inline-block;
+  ${(props) => css`
+    margin-right: ${props.align === 'grow' || props.align === 'start' ? 'auto' : undefined};
```
undefined는 왜주십니까?

KimChunsick replied:
허허.. 죄송함다..

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,118 @@
+    `}
+    ${props.shape &&
+    css`
+      width: ${props.width ? `${props.width}px` : undefined};
```
undefined를 왜쓰십니까?

KimChunsick replied:
허허.. 죄송함다..

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,118 @@
+}
+
+const Element = styled.button<ReturningUseButton>`
+  transition-property: border-color, background, color, transform, box-shadow;
```
transition-property가 왜 제일 위에 있습니까?

KimChunsick replied:
transition에 영향을 받는 내용들을 먼저 보여주고 밑에 transition에 대한 정보를 적어주는게 정보의 흐름을 따라가기 쉽지 않을까 생각해서 그랬습니다.

Vallista commented on packages/core/src/components/Button/type.ts:
```diff
@@ -0,0 +1,396 @@
+  prefix: React.ReactNode
+  suffix: React.ReactNode
+  color: 'primary' | 'secondary' | 'success' | 'error' | 'warning' | 'alert' | 'violet'
+  normal: { foreground: string; background: string; border: string }
```
`ThemeProvider`에 보면 `AvailablePickedColor` 가 있습니다.

Vallista commented on packages/core/src/components/Button/type.ts:
```diff
@@ -0,0 +1,396 @@
+  // Element: React.ElementType
+}
+
+const defaultVariantColorSets = {
```
타입추론으로 하면 문제가 있지 않을까요?

```tsx
type ButtonColorType = 'primary' | 'secondary' | 'success' | 'error' | 'warning' | 'alert' | 'violet'
type ButtonActionType = 'nomal' | 'hover' | 'action'
type ButtonVariantColorSets = Record<ButtonColorType, Record<ButtonActionType, AvailablePickedType>>

const defaultVariantColorSets: ButtonVariantColorSets = { ... }
```

1. 왜 이렇게 하는게 더 좋을까요?
2. 이 상황은 왜 타입추론을 하면 안될까요?

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,118 @@
+  const { children, prefix, suffix, disabled, loading, onClick, ...otherProps } = useButton(props)
+
+  return (
+    <Element type='submit' disabled={disabled || loading} {...otherProps} onClick={onClick}>
```
`submit` 은 고정인가요?

1. `'button' | 'submit' | 'reset'` 각각 지원을 해야할텐데, 왜 하지 않으셨나요?
2. sementic web을 전혀 지키지 않은 코드입니다. form과 연결하면 무조껀 button이 실행되겠군요?
3. submit은 variant와 color가 어떤 타입일 때 실행되어야 할까요? UI 적으로 생각해봅시다

KimChunsick replied:
vercel 에서 그렇게 했길래 그랬습니다... 수정하겠습니다.

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,118 @@
+    `}
+    box-shadow: ${props.variant === 'shadow' ? '0 5px 10px rgba(0,0,0,0.12)' : undefined};
+    color: ${props.normal?.foreground};
+    border: 1px solid ${props.normal?.border};
```
props.normal?.border 없으면 undefined.
그럼 `border: 1px solid undefined` 이므로 장애가 나겠군요..

KimChunsick replied:
해당 부분을 실수로 놓치고 커밋했네요. 수정하겠습니다.

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,118 @@
+      height: ${props.size === 'small' ? '32px' : props.size === 'medium' ? '40px' : '48px'};
+    `}
+    box-shadow: ${props.variant === 'shadow' ? '0 5px 10px rgba(0,0,0,0.12)' : undefined};
+    color: ${props.normal?.foreground};
```
props.normal?.foreground 없으면 undefined.
그럼 `color: 1px solid undefined` 이므로 장애가 나겠군요..

KimChunsick replied:
해당 부분을 실수로 놓치고 커밋했네요. 수정하겠습니다.

Vallista reviewed (changes requested):
근데 width 100% 하는 버튼일때는 어떻게 대응해요?

Vallista committed:
feat: Theme 옵션 추가

Commit:
fix: 1차 리뷰 반영

fix: style에서 삼항연산자로 undefined를 표현한 말도 안되는 코드 제거
fix: 하드코딩으로 박힌 box-shadow 값을 props.theme.shadows로 교체함
fix: 하드코딩 된 Color 값들을 Theme를 사용하여 불러오도록 수정
fix: 버튼 타입을 type prop으로 변경가능하게 수정

Signed-off-by: deokwon <deokwon@deokwonui-MacBookPro.local>

Vallista commented on packages/core/src/components/Button/useButton.ts:
```diff
@@ -0,0 +1,24 @@
+
+export function useButton<T extends Partial<ButtonProps>>(props: T): ReturningUseButton<T> {
+  const theme = useTheme()
+  const color = props.color || 'primary'
```
|| -> ??

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,152 @@
+      `};
+      ${props.variant === 'ghost' &&
+      css`
+        background-image: linear-gradient(to right, hsla(0, 0%, 100%, 0.8), hsla(0, 0%, 100%, 0.8));
```
^^

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,152 @@
+      `};
+      ${props.variant === 'ghost' &&
+      css`
+        background-image: linear-gradient(to right, hsla(0, 0%, 100%, 0.7), hsla(0, 0%, 100%, 0.7));
```
^^

Vallista commented on packages/core/src/components/Button/Button.tsx:
```diff
@@ -0,0 +1,152 @@
+    `}
+    ${props.variant === 'ghost' &&
+    css`
+      background-image: linear-gradient(to right, hsla(0, 0%, 100%, 0.8), hsla(0, 0%, 100%, 0.8));
```
^^

Vallista reviewed (changes requested):
^^

Vallista commented on packages/core/src/components/Button/Button.tsx:
````diff
@@ -0,0 +1,152 @@
+ * <Button type={Colors.HIGHLIGHT.PINK} size="small" variant="shadow" disabled>Hello World!</Button>
+ * ```
+ */
+// export const Button: FC<Partial<ButtonProps> & Partial<Pick<ButtonProps, 'normal' | 'hover' | 'active'>>> = (props) => {
````
주석

Commit:
fix: 2차 리뷰 반영

fix: transition style 위치 변경
fix: 쓸데없는 주석 제거

Signed-off-by: deokwon <deokwon@deokwonui-MacBookPro.local>

Vallista reviewed (approved):
ㅇㅋ
